### PR TITLE
applyset: introduce abstraction over objects

### PR DESCRIPTION
**Pull request by justinsb**
This lets us use typed and untyped objects.

### CODE REVIEW

**Comment by justinsb** [comment]
cc @apelisse

File: porch/controllers/remoterootsync/pkg/applyset/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+
+package applyset
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type ApplyableObject interface {
```
Comment (martinmaly) [inline porch/controllers/remoterootsync/pkg/applyset/interfaces.go:22]: This seems reasonable. I do wonder about some of the prior art interfaces that are similar but not quite the same. for example

[runtime.Object](https://github.com/kubernetes/apimachinery/blob/v0.23.6/pkg/runtime/interfaces.go#L302) has slightly different structure.

I prefer the straightforward simplicity of the ApplyableObject but it is worth checking whether there is a reason to go with some prior art already in place.

**Review by martinmaly — APPROVED** [review]

File: porch/controllers/remoterootsync/pkg/applyset/interfaces.go (hunk 1 above)
Comment (apelisse) [reply porch/controllers/remoterootsync/pkg/applyset/interfaces.go:22]: Yeah, we typically get this information with the `meta.Accessor` from the `runtime.Object`

Comment (justinsb) [reply porch/controllers/remoterootsync/pkg/applyset/interfaces.go:22]: Indeed, but the intent is that this isn't necessarily a runtime.Object.  It is supposed to be easy to take an unstructured object or a typed object and use it though!

**Pull request merged**